File: julia/benchmark_StrongDis_MM.jl
using BenchmarkTools

include("io.jl")
include("StrongDis.jl")

files = readdir("instances", join=true)

times = NamedTuple[]

mkpath("results")

for file in files

    if !endswith(file, ".csv")
        continue
    end

    println("Benchmarking $file")

    A, a, b = load_instance(file)

    # Warmup (JIT compilation)
    StrongDis_MM(A, a, b)

    # Eigentliche Messung
    start = time_ns()

    r = StrongDis_MM(A, a, b)

    elapsed = time_ns() - start

    push!(times, (
        file=file,
        time=elapsed,
        result=r
    ))
end

open("results/julia_results_MM.csv", "w") do io

    println(io, "file,time,result")

    for t in times
        println(io, "$(t.file),$(t.time),$(t.result)")
    end
end

File: rust/src/bin/benchmark_strong_dis_mm.rs
use std::fs::{self, File};
use std::io::{self, Write};
use std::path::Path;
use std::time::Instant;

use strongdis_runtime::io::{csv_instance_files, load_instance};
use strongdis_runtime::strong_dis_opt::strong_dis_mm_opt;

fn main() -> io::Result<()> {
    let repo_root = Path::new(env!("CARGO_MANIFEST_DIR")).join("..");
    let files = csv_instance_files(repo_root.join("instances"))?;
    let results_dir = repo_root.join("results");
    fs::create_dir_all(&results_dir)?;

    let mut rows = Vec::new();
    for file in files {
        println!("Benchmarking {}", file.display());
        let instance = load_instance(&file)?;

        strong_dis_mm_opt(&instance.matrix, instance.n, instance.a, instance.b);

        let start = Instant::now();
        let result = strong_dis_mm_opt(&instance.matrix, instance.n, instance.a, instance.b);
        let elapsed = start.elapsed().as_nanos();

        rows.push((file.display().to_string(), elapsed, result));
    }

    let mut output = File::create(results_dir.join("rust_results_MM_opt.csv"))?;
    writeln!(output, "file,time,result")?;
    for (file, time, result) in rows {
        writeln!(output, "{file},{time},{result}")?;
    }

    Ok(())
}
